src/transactions/test/TxResultsTests.cpp

```diff
@@ -113,6 +112,236 @@ TEST_CASE("txresults", "[tx][txresults]")
         return ltx.loadHeader().current().scpValue.closeTime;
     };
 
+    // set up world
+    auto root = TestAccount::createRoot(*app);
+    auto nonExisting = TestAccount{*app, getAccount("nonExisting")};
+    auto nonEnoughBalance =
+        root.create("nonEnoughBalance", lm.getLastMinBalance(0));
+    auto enoughBalance = root.create("enoughBalance", startAmount);
+    auto extraSignature = TestAccount{*app, getAccount("extraSignature")};
+
+    auto validateError = [&](TestAccount const& account,
+                             Transaction const& rawTx, int authLevel,
+                             ValidationResult const& validationResult,
+                             TransactionResult const& applyResult = {}) {
+        auto tx = account.unsignedTx(rawTx);
+
+        for_all_versions(*app, [&] {
+            auto version = [&]() {
+                LedgerTxn ltx(app->getLedgerTxnRoot());
+                return ltx.loadHeader().current().ledgerVersion;
+            }();
+
+            SECTION("no signatures")
+            {
+                if (version == 7 || authLevel < 1)
+                {
+                    validateTxResults(tx, *app, validationResult, applyResult);
+                }
+                else
+                {
+                    validateTxResults(tx, *app,
+                                      {validationResult.fee, txBAD_AUTH});
+                }
+            }
+
+            SECTION("one signature")
+            {
+                account.sign(tx);
+                validateTxResults(tx, *app, validationResult, applyResult);
+            }
+
+            SECTION("double signature")
+            {
+                account.sign(tx);
+                account.sign(tx);
+
+                if (version == 7 || authLevel < 2)
+                {
+                    validateTxResults(tx, *app, validationResult, applyResult);
+                }
+                else
+                {
+                    validateTxResults(tx, *app,
+                                      {validationResult.fee, txBAD_AUTH_EXTRA});
+                }
+            }
+
+            SECTION("extra signature")
+            {
+                account.sign(tx);
+                extraSignature.sign(tx);
+
+                if (version == 7 || authLevel < 2)
+                {
+                    validateTxResults(tx, *app, validationResult, applyResult);
+                }
+                else
+                {
+                    validateTxResults(tx, *app,
+                                      {validationResult.fee, txBAD_AUTH_EXTRA});
+                }
+            }
+        });
+    };
+
+    SECTION("transaction errors")
+    {
+        SECTION("non existing")
+        {
+            SECTION("all possible bugs")
```

Comment:
"all possible bugs" doesn't mean anything: what you're testing is `txMISSING_OPERATION`